dns/message.py

```diff
@@ -488,11 +489,33 @@ def to_wire(
         *tsig_ctx*, a ``dns.tsig.HMACTSig`` or ``dns.tsig.GSSTSig`` object, the
         ongoing TSIG context, used when signing zone transfers.
 
+        *padding*, an ``int`` or ``None``.  If EDNS is in use and padding is
+        not ``None``, then the message is padded to a multiple of the specified
+        value if > 0, or to the RFC 8467 recommended default value if 0.
+
         Raises ``dns.exception.TooBig`` if *max_size* was exceeded.
 
         Returns a ``bytes``.
         """
 
+        size_without_padding = 0
+        if self.opt and padding is not None and len(self.additional) > 0:
+            # We have additional data but won't know how much padding we need when
+            # we render the opt, so figure out how big a padding-less message is.
+            # It's unfortunate we have to do this instead of adding something
+            # to the end.
+            w = self.to_wire(origin, max_size, multi, tsig_ctx, padding=None, **kw)
+            size_without_padding = len(w)
+            if size_without_padding >= 2**14:
+                # This is a big message so we just punt and don't pad.  Doing "the
+                # right thing" and padding here is kind of complicated because of
+                # potential compression in the additional data section.  Adding padding
+                # based on this size might push something compressed into the
+                # incompressible zone, and our size computation will then be wrong.
+                # If we're really unlucky, it will also cause truncation!  Solving this
+                # could involve multiple re-renders and/or selectively deleting
+                # additional data.
+                return w
```

Comment:
This seems to be overly complicated by the fact that the OPT record is rendered before other records in the additional section.  Rendering the OPT record after these other records would remove much of the need for this (there's still some complication with TSIG, but I think that can be handled differently)

I also think this might have (several) problems if TSIG is in use.  First, it's checking `len(self.additional)` to see if there are additional records, but it's not checking `tsig_ctx`, so it's not going to catch the case where there will be a TSIG record after the OPT, but no other additional records.  Second, if `multi` is true, the TSIG context is being updated in place, and when `to_wire` is called multiple times for the same message, the non-idempotent update will be done twice.